cmd/docker-proxy/main.go

```diff
@@ -13,11 +13,27 @@ import (
 	"github.com/ishidawataru/sctp"
 )
 
+// The caller is expected to pass-in open file descriptors ...
+const (
+	// Pipe for reporting status, as a string. "0\n" if the proxy
+	// started normally. "1\n<error message>" otherwise.
+	parentPipeFd uintptr = 3 + iota
+	// If -use-listen-fd=true, a listening socket ready to accept TCP
+	// connections or receive UDP. (Without that option on the command
+	// line, the listener needs to be opened by docker-proxy, for
+	// compatibility with older docker daemons. In this case fd 4
+	// may belong to the Go runtime.)
+	listenSockFd
+)
+
 func main() {
```

Comment:
Since we're thinking about close-on-exec already: https://github.com/golang/go/blob/1985c0ccf9ef0736aeb79ef548689aa935fa5c4c/src/syscall/exec_linux.go#L611-L612

I know that docker-proxy does not exec, and probably won't in the foreseeable future, but passing fd 3 or 4 to a child process would result in such subtle lifecycle bugs (e.g. parent pipe isn't closed until grandchild exits, listen socket remains bound until grandchild exits) that I feel it's in our best interest to preemptively fix it.

```suggestion
func main() {
	// Mark any files we expect to inherit as close-on-exec
	// so that they are not unexpectedly inherited by any child processes
	// if we ever need docker-proxy to exec something.
	// This is safe to do even if the fd belongs to the Go runtime
	// as it would be a no-op:
	// the Go runtime marks all file descriptors it opens as close-on-exec.
	// See the godoc for syscall.ForkLock for more information.
	syscall.CloseOnExec(parentPipeFd)
	syscall.CloseOnExec(listenSockFd)
```

Reply:
Ok